converters/fromZigbee.js

```diff
@@ -2585,13 +2585,13 @@ const converters = {
                 result[postfixWithEndpointName('trigger_time', msg, model)] = msg.data[0x4011];
             }
             if (typeof msg.data[0x4012] == 'number') {
-                result[postfixWithEndpointName('mounted_mode', msg, model)] = (msg.data[0x4012]==1);
+                result[postfixWithEndpointName('mounted_mode_active', msg, model)] = (msg.data[0x4012]==1);
```

Comment:
Might be worth to add those cluster attributes here:
https://github.com/Koenkk/zigbee-herdsman/blob/ef3d96c9ed4b75c57af1fc71e618add0f1c3064e/src/zcl/definition/cluster.ts#L1899 it should improve readability a lot.

Line I linked was for the Viessmann which is a similar TRV but without any documentation.

Reply:
@sjorge Initially I have considered this https://github.com/robertalexa/zigbee-herdsman-converters/commit/dfccce4774f6356560927363be55d537aabcd6cd

However after discussing this with @Koenkk as part of a different issue he has advised agaisnt (unless i am misunderstanding this fully)

See issue here https://github.com/Koenkk/zigbee2mqtt/issues/7065

The fact that those clusters are named "CustomX" isn't great either, but based on the Danfoss specs (i too added support for Hive kinda blindly) I can map what some of those customs actually are.

viessmannCustom0: eTRV Open Window Detection - window_open_internal
viessmannCustom1: not sure
viessmannCustom2: not sure
viessmannWindowOpenForce: External Open Window Detected - window_open_external
viessmannCustom4: not sure
viessmannCustom5: not sure
viessmannCustom6: not sure
viessmannCustom7: Exercise day of week
viessmannCustom8: Exercise trigger time 
viessmannAssemblyMode: Mounting mode active
viessmannCustom10: Control algorithm scale factor
viessmannCustom11: not sure
viessmannCustom12: Regulation SetPoint Offset
viessmannCustom13: Adaptation run control
viessmannCustom14: Adaptation run status
viessmannCustom15: Adaptation run settings

Maybe in future we get them renamed.

@Koenkk Do you think the Danfoss Ally and Hive TRV should use clusters or just stick with msg.data[address]?

Rob

Reply:
Let's see what Koen says about it. If he suggests using clusters instead of attribute address, then I can go through them, remove unused ones, rename the remaining and fix the devices depending on them as part of this PR.

But i agree that some of them should go, especially since we don not know what they actually do.

Rob

Reply:
Yeah I labeled them all and went through days of logging but they not all the same as the danfoss ones. They use a diff manufacture too.

Viessman was not helpful ar all when contact and they told me to stop reverse engineering there device.

Reply:
LOL :) Typical corporate greed I suppose.

I contacted Danfoss to ask them if and how to flash their firmware on the Hive TRV (i dont like Hive, learnt the hard way) - they replied with a generic message...

I contacted Hive to ask about their specs, completely oblivious. They confirmed the device is identical to Danfoss Ally but is using a custom firmware, so I have no idea if it supports more or less features, but so far everything matched.

I do have some issues with thermostat_orientation though. The command to get and set work based on Danfoss Ally specs, but the display is not turning. Meh. But I have left it in so Danfoss users will benefit from it.

Reply:
It's fine to add these to cluster.ts file, as long as the attributes have a manufacturerCode like the Viesmann ones.

Reply:
@sjorge Did you say you already have something started on this or should I handle it, including "remove unused ones, rename the remaining and fix the devices depending on them as part of this PR."

@Koenkk Are you ok with me handling all of this in one go? Do another PR on herdsman and when ready we re-review this one?

Rob

Reply:
> @sjorge Did you say you already have something started on this or should I handle it, including "remove unused ones, rename the remaining and fix the devices depending on them as part of this PR."

Working on the PR now, nearly done.
It's basically rename viessmannCustom0 -> viessmanWindowOpen and drop the other viessmannCustomX's.
And update the reference to viessmannCustom0 in zhc ofcourse.

Reply:
Is it worth renaming that to viessmanWindowOpenInternal? Because some trvs have 2 one for internal and one for external, like the ones at hand.

I will wait for your PRs to be merged, then start with "good" code basically and deal with my clusters.

Thanks for confirming

Reply:
Maybe? They Viessmann seems to be the same as the Danfoss with a very buggy firmware.

Reply:
Seems plausible, don’t have the TRV paired atm to verify.

Reply:
@sjorge As part of me adding the clusters for the Danfoss, do you mind if I rename (herdsman and converters) viessmannWindowOpenForce to viessmannWindowOpenExternal? Thanks

Reply:
@robertalexa  you probably want to rename the attribute names in the mqtt payload too then.
I did look at the existing danfoss stuff and it didn't quiet line up, so aside the naming, the code should probably stay the same to be sure.

I don't have access to the viessmann to retest.

Reply:
@sjorge I shall leave it as is. I guess it won't harm anyone. Any realistically we don't really know if they actually call it external. Thanks for confirming anyway

Reply:
@sjorge Please help. At this point I can't tell if I am tired, stupid or just missing something. As you know there is s PR for herdsman clusters.
toZigbee.js
This code (old) works:
```
danfoss_mounted_mode_control: {
        key: ['mounted_mode_control'],
        convertSet: async (entity, key, value, meta) => {
            const payload = {0x4013: {value: (value ? 0x00: 0x01), type: 0x10}};
            await entity.write('hvacThermostat', payload, manufacturerOptions.danfoss);
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('hvacThermostat', [0x4013], manufacturerOptions.danfoss);
        },
    }
```
This code (new) does not:
```
danfoss_mounted_mode_control: {
        key: ['mounted_mode_control'],
        convertSet: async (entity, key, value, meta) => {
            const payload = {'danfossMountedModeControl': {value: (value ? 0x00: 0x01), type: 0x10}};
            await entity.write('hvacThermostat', payload, manufacturerOptions.danfoss);
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('hvacThermostat', ['danfossMountedModeControl'], manufacturerOptions.danfoss);
        },
    }
```

What am I missing? Or what other info can i send you for debugging? Thanks in advance.

fromZigbee.js
```
if (msg.data.hasOwnProperty('danfossMountedModeControl')) {
                result[postfixWithEndpointName('mounted_mode_control', msg, model)] = (msg.data['danfossMountedModeControl'] === 0);
            }
```
Getting the info works fine, setting it doesn't

Also here is some more debug info for both, to compare
```
Zigbee2MQTT:debug 2021-05-20 22:25:55: Publishing 'set' 'mounted_mode_control' to '0x842e14fffe203c4c'
  zigbee-herdsman:controller:endpoint Write 0x842e14fffe203c4c/1 hvacThermostat({"danfossMountedModeControl":{"value":1,"type":16}}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4678,"transactionSequenceNumber":null,"writeUndiv":false}) +5s


Zigbee2MQTT:debug 2021-05-20 22:29:09: Publishing 'set' 'mounted_mode_control' to '0x842e14fffe203c4c'
  zigbee-herdsman:controller:endpoint Write 0x842e14fffe203c4c/1 hvacThermostat({"16403":{"value":1,"type":16}}, {"sendWhenActive":false,"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4678,"transactionSequenceNumber":null,"writeUndiv":false}) +0ms
```
This is happening for SET on all the custom commands
@Koenkk Any thoughts?

Reply:
I have commited my code so you can have an easy before/after
https://github.com/Koenkk/zigbee-herdsman-converters/pull/2592/commits/ea4a84aaaeb11b5b7e0f30945cbb7601bbc6ca0e

Hope this makes it easier.

Reply:
I think I know what is wrong, as the info is now in herdsman, you do not need to specify the data type.
```
const payload = {'danfossMountedModeControl': {value: (value ? 0x00: 0x01), type: 0x10}};
```
Should be
```
const payload = {'danfossMountedModeControl': (value ? 0x00: 0x01)};
```

Reply:
Thanks for pointing me in the right direction. You are absolutely right, but I initially didn't find it obvious. Thanks. I have pushed another commit which I hope will "Resolve this conversation"

I would appreciate some review and feedback in hive/danfoss.js

```
            // danfoss attributes
            await endpoint.configureReporting('hvacThermostat', [{attribute: 'danfossMountedModeActive', minimumReportInterval: 0,
                maximumReportInterval: constants.repInterval.MINUTES_10, reportableChange: 1}], options);
            await endpoint.configureReporting('hvacThermostat', [{attribute: 'danfossWindowOpenInternal', minimumReportInterval: 0,
                maximumReportInterval: constants.repInterval.HOUR, reportableChange: 1}], options);
            await endpoint.configureReporting('hvacThermostat', [{attribute: 'danfossHeatRequired', minimumReportInterval: 0,
                maximumReportInterval: constants.repInterval.MINUTES_10, reportableChange: 1}], options);

            // read keypadLockout, we don't need reporting as it cannot be set physically on the device
            await endpoint.read('hvacUserInterfaceCfg', ['keypadLockout']);
            await endpoint.read('hvacThermostat', [
                'danfossWindowOpenExternal',
                'danfossDayOfWeek',
                'danfossTriggerTime',
                'danfossAlgorithmScaleFactor',
                'danfossHeatAvailable',
                'danfossMountedModeControl'
            ], options);
```

I am not sure I did the right thing. This is the technical specs https://assets.danfoss.com/documents/176987/AM375549618098en-000101.pdf

But I would be very happy to be "schooled"

Thanks a lot @sjorge for your involvement and commitment with this PR and all the work around it.

Reply:
Mostly looks good, left a few comments that should make the code more readable and more conssitant.

Reply:
That pdf has some interesting stuff in it, looks like if you have multiple they can somehow be bound and tell eachother to enter pre-heat if I am reading that right... anyway before I get to distracted.

We might want to tweak the default reporting a bit to save on battery
```
            await reporting.batteryPercentageRemaining(endpoint, {min: 60, max: 43200, change: 1});
            await reporting.thermostatTemperature(endpoint, {min: 0, max: constants.repInterval.MINUTES_10, change: 25});
            await reporting.thermostatPIHeatingDemand(endpoint, {min: 0, max: constants.repInterval.MINUTES_10, change: 1});
            await reporting.thermostatOccupiedHeatingSetpoint(endpoint, {min: 0, max: constants.repInterval.MINUTES_10, change: 25});
```

Seems rather aggressive, we can probably bumb batteryPercentageRemaining to min=1h or maybe even 3h, I doubt the TRV will drop by 1% every minute.


thermostatTemperature, thermostatPIHeatingDemand, thermostatOccupiedHeatingSetpoint dont have a minumum either. Perhaps pushing those to min=15sec or maybe even min=60sec

Reply:
@sjorge I have changed the battery to 1 hours. With regards to the rest, correct me if I am wrong but if we change reporting of for example thermostatOccupiedHeatingSetpoint to 60 seconds, and then you change the temperature physically on the TRV, it can then take up to 60 seconds for it to be reported back to z2h?

Apart from the 3 you have mentioned there are these ones with a min of 0 too. Would you suggest we increase them too?
```
await endpoint.configureReporting('hvacThermostat', [{attribute: 'danfossMountedModeActive', minimumReportInterval: 0,
                maximumReportInterval: constants.repInterval.MINUTES_10, reportableChange: 1}], options);
            await endpoint.configureReporting('hvacThermostat', [{attribute: 'danfossWindowOpenInternal', minimumReportInterval: 0,
                maximumReportInterval: constants.repInterval.HOUR, reportableChange: 1}], options);
            await endpoint.configureReporting('hvacThermostat', [{attribute: 'danfossHeatRequired', minimumReportInterval: 0,
                maximumReportInterval: constants.repInterval.MINUTES_10, reportableChange: 1}], options);
```

Also would you be kind to explain to me what exactly does this read do? I must confess that I have filled it in but without full understanding of the behaviour
```
await endpoint.read('hvacThermostat', [
                    'danfossWindowOpenExternal',
                    'danfossDayOfWeek',
                    'danfossTriggerTime',
                    'danfossAlgorithmScaleFactor',
                    'danfossHeatAvailable',
                    'danfossMountedModeControl'
                ], options);
```

Thanks, Rob

Reply:
> example thermostatOccupiedHeatingSetpoint to 60 seconds, and then you change the temperature physically on the TRV, it can then take up to 60 seconds for it to be reported back to z2h?

Sort of ...
- physically set the temp to 24 on the TRV
- report gets send
- physically set the temp to something else within minInterval
- report gets delayed until minInterval is passed

We can safely raise the minimum for anything that is not physically changeable on the device.

> Also would you be kind to explain to me what exactly does this read do? I must confess that I have filled it in but without full understanding of the behaviour

To get the initial state we read some attributes, usually for things that do not have reporting setup and are not user trigger-able on the device. e.g. for the viessmann we read the keypadLockout from hvacUserInterface, so we have a value available for the frontend. They only way to set it, is via mqtt messages so it makes no sense to setup reporting for that attribute as it cannot be changed without going through zigbee2mqtt.

Reply:
Based on your reply I have done the following:
https://github.com/Koenkk/zigbee-herdsman-converters/pull/2592/commits/cf0c6b41fa8272477b7c6493807523a44e3952c7

So the 1st reporting will be instant as a result of a change and the 2nd report will be delayed up to minInterval from 1st report.

I hope this change will "Resolve this conversation" chain. But I will leave this up to you to close? (i hope you can see the button too?)

Reply:
Left one more comment, but will resolve this one as that one shows up as a seperate thread.